File: internal/httpapi/login_audits_test.go
package httpapi

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/gogomail/gogomail/internal/admin"
)

type mockLoginAuditService struct {
	lastFilter admin.LoginAuditFilter
	logs       []admin.LoginAuditLog
}

func (m *mockLoginAuditService) ListLoginAttempts(ctx context.Context, filter admin.LoginAuditFilter) ([]admin.LoginAuditLog, error) {
	m.lastFilter = filter
	return m.logs, nil
}

func TestCompanyLoginAuditList(t *testing.T) {
	mockSvc := &mockLoginAuditService{
		logs: []admin.LoginAuditLog{
			{
				ID:            "login-1",
				UserID:        "user-1",
				CompanyID:     "company-1",
				IPAddress:     "127.0.0.1",
				UserAgent:     "Mozilla/5.0",
				Success:       false,
				FailureReason: "bad password",
				Timestamp:     time.Date(2026, time.May, 15, 10, 0, 0, 0, time.UTC),
			},
		},
	}

	mux := http.NewServeMux()
	mux.HandleFunc("GET /admin/v1/companies/{id}/security/login-audits", handleCompanyLoginAudits(mockSvc))

	server := httptest.NewServer(mux)
	defer server.Close()

	resp, err := http.Get(server.URL + "/admin/v1/companies/company-1/security/login-audits?user_id=user-1&success=false&limit=10&offset=5&from_date=2026-05-01T00:00:00Z&to_date=2026-05-31T23:59:59Z")
	if err != nil {
		t.Fatal(err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		t.Fatalf("status = %d", resp.StatusCode)
	}

	var body struct {
		LoginAudits []struct {
			ID            string `json:"id"`
			UserID        string `json:"user_id"`
			CompanyID     string `json:"company_id"`
			IPAddress     string `json:"ip_address"`
			UserAgent     string `json:"user_agent"`
			Success       bool   `json:"success"`
			FailureReason string `json:"failure_reason"`
			Timestamp     string `json:"timestamp"`
		} `json:"login_audits"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&body); err != nil {
		t.Fatal(err)
	}
	if len(body.LoginAudits) != 1 || body.LoginAudits[0].ID != "login-1" {
		t.Fatalf("body = %+v", body)
	}
	if mockSvc.lastFilter.CompanyID != "company-1" || mockSvc.lastFilter.UserID != "user-1" || mockSvc.lastFilter.Success == nil || *mockSvc.lastFilter.Success {
		t.Fatalf("filter = %+v", mockSvc.lastFilter)
	}
	if mockSvc.lastFilter.Limit != 10 || mockSvc.lastFilter.Offset != 5 {
		t.Fatalf("pagination = %+v", mockSvc.lastFilter)
	}
	if mockSvc.lastFilter.StartTime == nil || mockSvc.lastFilter.EndTime == nil {
		t.Fatalf("time filters = %+v", mockSvc.lastFilter)
	}
}

File: apps/console/src/lib/__tests__/loginAudits.test.ts
import { describe, expect, it } from 'vitest';

import { buildLoginAuditsQuery, exportLoginAuditsCsv } from '../loginAudits';

describe('loginAudits', () => {
  it('builds a company-scoped login audit query', () => {
    expect(
      buildLoginAuditsQuery({
        companyId: ' company-1 ',
        userId: ' user-7 ',
        success: false,
        fromDate: '2026-05-01T00:00:00Z',
        toDate: '2026-05-31T23:59:59Z',
        limit: 100,
        offset: 25,
      })
    ).toBe(
      'company_id=company-1&user_id=user-7&success=false&from_date=2026-05-01T00%3A00%3A00Z&to_date=2026-05-31T23%3A59%3A59Z&limit=100&offset=25'
    );
  });

  it('exports login audit rows as escaped csv', () => {
    expect(
      exportLoginAuditsCsv([
        {
          id: 'login-1',
          user_id: 'user-1',
          company_id: 'company-1',
          ip_address: '127.0.0.1',
          user_agent: 'Mozilla/5.0',
          success: false,
          failure_reason: 'bad password',
          timestamp: '2026-05-15T10:00:00Z',
        },
        {
          id: 'login-2',
          user_id: 'user-2',
          company_id: 'company-1',
          ip_address: '',
          user_agent: 'curl/8.0',
          success: true,
          failure_reason: '',
          timestamp: '2026-05-15T10:30:00Z',
        },
      ])
    ).toBe(
      [
        'id,user_id,company_id,ip_address,user_agent,success,failure_reason,timestamp',
        '"login-1","user-1","company-1","127.0.0.1","Mozilla/5.0","false","bad password","2026-05-15T10:00:00Z"',
        '"login-2","user-2","company-1","","curl/8.0","true","","2026-05-15T10:30:00Z"',
      ].join('\n')
    );
  });
});
